File: code/library/src/main/java/org/fbme/lib/iec61499/parser/DeviceTypeConverter.java
package org.fbme.lib.iec61499.parser;


import org.fbme.lib.common.Identifier;
import org.fbme.lib.iec61499.declarations.DeviceTypeDeclaration;
import org.fbme.lib.iec61499.declarations.ResourceDeclaration;
import org.fbme.lib.iec61499.declarations.ResourceTypeReference;
import org.jdom.Element;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class DeviceTypeConverter extends DeclarationConverterBase<DeviceTypeDeclaration> {

    public DeviceTypeConverter(ConverterArguments arguments) {
        super(arguments);
    }

    @Override
    protected DeviceTypeDeclaration extractDeclarationBody(@Nullable Identifier identifier) {
        assert myElement != null;

        DeviceTypeDeclaration dtd = myFactory.createDeviceTypeDeclaration(identifier);
        ParameterDeclarationConverter.extractAll(this, dtd.getParameters());

        List<Element> resoruceTypeNameElements = myElement.getChildren("ResourceTypeName");
        for (Element resourceTypeNameElement : resoruceTypeNameElements) {
            ResourceTypeReference resourceTypeReference = myFactory.createResourceTypeReference();
            resourceTypeReference.getResourceType().setTargetName(resourceTypeNameElement.getAttributeValue("Name"));
            dtd.getInstantiableResourceTypes().add(resourceTypeReference);
        }

        for (Element resourceElement : myElement.getChildren("Resource")) {
            ResourceDeclaration resource = new ResourceConverter(with(resourceElement)).extract();
            dtd.getResources().add(resource);
        }

        Element deviceFBNetwork = myElement.getChild("FBNetwork");
        if (deviceFBNetwork != null) {
            new FBNetworkConverter(with(deviceFBNetwork.getChild("FBNetwork")), dtd.getNetwork()).extractNetwork();
        }

        return dtd;
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/parser/DeviceTypeConverter.kt
package org.fbme.lib.iec61499.parser

import org.fbme.lib.common.Identifier
import org.fbme.lib.iec61499.declarations.DeviceTypeDeclaration

class DeviceTypeConverter(arguments: ConverterArguments) :
    DeclarationConverterBase<DeviceTypeDeclaration>(arguments) {
    override fun extractDeclarationBody(identifier: Identifier?): DeviceTypeDeclaration {
        checkNotNull(element)
        val dtd = factory.createDeviceTypeDeclaration(identifier)
        ParameterDeclarationConverter.extractAll(this, dtd.parameters)
        val resoruceTypeNameElements = element.getChildren("ResourceTypeName")
        for (resourceTypeNameElement in resoruceTypeNameElements) {
            val resourceTypeReference = factory.createResourceTypeReference()
            resourceTypeReference.resourceType.setTargetName(resourceTypeNameElement.getAttributeValue("Name"))
            dtd.instantiableResourceTypes.add(resourceTypeReference)
        }
        for (resourceElement in element.getChildren("Resource")) {
            val resource = ResourceConverter(with(resourceElement)).extract()
            dtd.resources.add(resource)
        }
        val deviceFBNetwork = element.getChild("FBNetwork")
        if (deviceFBNetwork != null) {
            FBNetworkConverter(with(deviceFBNetwork.getChild("FBNetwork")), dtd.network).extractNetwork()
        }
        return dtd
    }
}
